codebase/src/cpp/ieee1516e/src/portico/DatatypeRetrieval.cpp

```diff
@@ -0,0 +1,320 @@
+#include "DatatypeRetrieval.h"
+#include "utils\StringUtils.h"
+
+PORTICO1516E_NS_START
+
+const string DatatypeRetrieval::BASIC = "basicData";
+const string DatatypeRetrieval::SIMPLE = "simpleData";
+const string DatatypeRetrieval::ENUMERATED = "enumeratedData";
+const string DatatypeRetrieval::ARRAY = "arrayData";
+const string DatatypeRetrieval::FIXED = "fixedRecordData";
+const string DatatypeRetrieval::VARIANT = "variantRecordData";
+const string DatatypeRetrieval::NA = "NA";
+
+DatatypeRetrieval::DatatypeRetrieval()
+{
+	// We store this so we can decide if we want to update 
+	// the FOM at runtime. 
+	this->initialized = false;
+}
+
+DatatypeRetrieval::~DatatypeRetrieval()
+{}
+
+void DatatypeRetrieval::initialize(std::wstring fomString)
+{
+	string xmlIn = pugi::as_utf8(fomString);
+	pugi::xml_parse_result result = this->fomxml.load_string(xmlIn.c_str());
+
+	// The FOM was parsed set initialized ot true.
+	if (result.status == pugi::xml_parse_status::status_ok)
+	{
+		this->initialized = true;
+	}
+	else{
+		// throw error that FOM could not be parsed
+		throw RTIinternalError(L"The FOM could not be parsed. Please check FOM is correct.");
+	}
+}
+
+bool DatatypeRetrieval::isInitialized()
+{
+	return this->initialized;
+}
+
+IDatatype* DatatypeRetrieval::getParameterDatatype(string dataTypeName)
+{
+	return getDatatype(dataTypeName);
+}
+
+IDatatype* DatatypeRetrieval::getAttributeDatatype(string dataTypeName)
+{
+	return getDatatype(dataTypeName);
+}
+
+IDatatype* DatatypeRetrieval::getDatatype(string dataTypeName)
+{
+	IDatatype* datatype = nullptr;
+
+	// If it hasn't been cached then create it and cache it.
+	if (this->typeCache.find(dataTypeName) == this->typeCache.end())
+	{
+		if (dataTypeName == "NA")
+		{
+			datatype = new NaType();
+		}
+		else
+		{
+			// create and cache it
+			pugi::xml_node typeNode = getDatatypeNode(dataTypeName);
+			string classTypeName = typeNode.name();
+
+			if (classTypeName == DatatypeRetrieval::BASIC)
+			{
+				datatype = getBasicType(typeNode);
+			}
+			else if (classTypeName == DatatypeRetrieval::SIMPLE)
+			{
+				datatype = getSimpleType(typeNode);
+			}
+			else if (classTypeName == DatatypeRetrieval::ENUMERATED)
+			{
+				datatype = getEnumeratedType(typeNode);
+			}
+			else if (classTypeName == DatatypeRetrieval::ARRAY)
+			{
+				datatype = getArrayType(typeNode);
+			}
+			else if (classTypeName == DatatypeRetrieval::FIXED)
+			{
+				datatype = getFixedRecordType(typeNode);
+			}
+			else if (classTypeName == DatatypeRetrieval::VARIANT)
+			{
+				datatype = getVariantRecordType(typeNode);
+			}
+		}
+
+		// cache it
+		this->typeCache[dataTypeName] = datatype;
+	}
+	else
+	{
+		// Get from cache using handle as key. 
+		datatype = this->typeCache[dataTypeName];
+	}
+
+	return datatype;
+}
+
+
+
+
+IDatatype* DatatypeRetrieval::getBasicType(pugi::xml_node dataNode)
+{
+
+	// Get the parameters from the node
+	string typeName = dataNode.attribute("name").as_string();
```

Comment:
This is set on the Java side when creating the XML structure to pass to the cpp. I think that it will be safe, or throw error on the java side if there is a problem.